File: Day233_2025/Q233_GFG.cpp
class Solution {
    bool isPossible(vector<int>& arr, int gap, int k){
        int count = 1;
        int prev = arr[0];
        for(int i = 1; i < arr.size(); i++)
        {
            if( arr[i] - prev >= gap)
            {
                count++;
                prev = arr[i];
            }
            if(count == k)
                return true;
        }
        return false;
    }
  public:
    int maxMinDiff(vector<int>& arr, int k) {
        // Q. Maximize the minimum difference between k elements
        // code here
        int n = arr.size();
        sort(arr.begin(),arr.end());
        int low  = 0;
        int high = arr[n-1] - arr[0];
        int ans = 0;
        while(low <= high)
        {
            int mid = (low + high)/2;
            if(isPossible(arr,mid,k))
            {
                ans = mid;
                low = mid + 1;
            }
            else
            {
                high = mid-1;
            }
        }
        return ans;
    }
};

File: Day233_2025/Q233_GFG.py
class Solution:
    def isPossible(self, arr, gap, k):
        count = 1
        prev = arr[0]
        for i in range(1, len(arr)):
            if arr[i] - prev >= gap:
                count += 1
                prev = arr[i]
            if count == k:
                return True
        return False

    def maxMinDiff(self, arr, k):
        arr.sort()
        low, high = 0, arr[-1] - arr[0]
        ans = 0

        while low <= high:
            mid = (low + high) // 2
            if self.isPossible(arr, mid, k):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1
        return ans
